pkg/controller/provider/xavier/base/client.go

```diff
@@ -0,0 +1,335 @@
+					recommended = true
+					break
+				}
+			}
+		}
+		if !recommended {
+			concerns = append(
+				concerns,
+				"CNV not recommended.")
+		}
+	}
+	r.Adapter.UpdateConcerns(model, concerns)
+
+	Log.Info("Analyzed", "vm", model.String())
+
+	return
+}
+
+//
+// Build body
+func (r *Client) body(model libmodel.Model) (body Object) {
+	about, _ := r.about()
+	object := Object{
+		Provider:        "",
+		DataCenter:      "",
+		Cluster:         "",
+		VmName:          "",
+		DiskSpace:       0,
+		Memory:          0,
+		CpuCores:        0,
+		GuestOs:         "",
+		OsProduct:       "",
+		Product:         about.Product,
+		Version:         about.APIVersion,
+		HostName:        "",
+		HasMemoryHotAdd: false,
+		HasCpuHotAdd:    false,
+		HasCpuHotRemove: false,
+		HasCpuAffinity:  false,
+		HasDrsEnabled:   false,
+		HasHaEnabled:    false,
+		HasPassthrough:  false,
+		HasUsb:          false,
+		HasRdmDisk:      false,
+		BalloonedMemory: 0,
+		ScanDate:        time.Now().Unix(),
+	}
+	insert := Object{
+		"object": Object{
+			"org.jboss.xavier.analytics.pojo.input.workload.inventory.VMWorkloadInventoryModel": object,
+		},
+		"out-identifier": "input",
+	}
+	body = Object{
+		"lookup": "WorkloadInventoryKSession0",
+		"commands": List{
+			Object{
+				"insert": insert,
+			},
+			Object{
+				"fire-all-rules": Object{
+					"out-identifier": "firedActivations",
+				},
+			},
+			Object{
+				"query": Object{
+					"name":           "GetWorkloadInventoryReports",
+					"out-identifier": "WorkloadInventoryReports",
+					"arguments":      List{},
+				},
+			},
+		},
+	}
+	r.Adapter.UpdateBody(object, model)
+
+	return
+}
+
+//
+// Add headers.
+func (r Client) header() (header http.Header) {
+	header = http.Header{}
+	header.Add("Content-Type", "application/json")
+	encoded := base64.StdEncoding.EncodeToString(
+		[]byte(strings.Join([]string{
+			r.User,
+			r.Password,
+		}, ":")))
+	header.Add(
+		"Authorization",
+		"Basic "+encoded)
+
+	return
+}
+
+//
+// Post validation request.
+func (r *Client) post(in, out interface{}) (status int, err error) {
+	parsedURL, err := liburl.Parse(r.URL)
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	body, _ := json.Marshal(in)
+	reader := bytes.NewReader(body)
+	request := &http.Request{
+		Method: http.MethodPost,
+		Body:   ioutil.NopCloser(reader),
+		URL:    parsedURL,
+		Header: r.header(),
+	}
+	client := http.Client{}
+	response, err := client.Do(request)
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	status = response.StatusCode
+	content := []byte{}
+	if status == http.StatusOK {
+		defer response.Body.Close()
+		content, err = ioutil.ReadAll(response.Body)
+		if err != nil {
+			err = liberr.Wrap(err)
+			return
+		}
+		err = json.Unmarshal(content, out)
+		if err != nil {
+			err = liberr.Wrap(err)
+			return
+		}
+	}
+
+	return
+}
+
+//
+// Get object at path.
+func (r Client) find(in interface{}, path ...string) (out interface{}, err error) {
```

Comment:
Needed to safely traverse the absurd REST payloads.

Reply:
Likewise, `in` and `out` here both look like they are only ever a `base.Object`.

Reply:
Not really.  Depending on the path, the `out` could be a `List`, string or anything.   The `out` is used as `in` in other places so even for `in` to be typed as `*Object` would make things difficult.